### PR TITLE
Stay logged in to MSA after app restart

### PR DESCRIPTION
**Pull request by annakocheshkova**
Things to consider before you submit the PR:

* [X] Are tests passing locally?
* [X] Are the files formatted correctly?
* [X] Did you check UI tests on the sample app? They are not executed on CI.
* [X] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?

## Description

Persist MSA userId.

## Related PRs or issues

[AB#76752](https://msmobilecenter.visualstudio.com/454a20a9-dfe6-4c1f-8ae8-417f76adb472/_workitems/edit/76752)

### CODE REVIEW

File: Sasquatch/Sasquatch/ViewControllers/MSMainViewController.swift [hunk 1]
```diff
@@ -41,8 +41,11 @@ class MSMainViewController: UITableViewController, AppCenterProtocol {
   private var startupModePicker: MSEnumPicker<StartupMode>?
   private var eventFilterStarted = false
   private var dbFileDescriptor: CInt = 0
-  private var dbFileSource: DispatchSourceProtocol?  
+  private var dbFileSource: DispatchSourceProtocol?
+
   let startUpModeForCurrentSession: NSInteger = (UserDefaults.standard.object(forKey: kMSStartTargetKey) ?? 0) as! NSInteger
+  private let kMSAUserIdKey = "MSAUserId"
```
Comment (Jamminroot) [inline Sasquatch/Sasquatch/ViewControllers/MSMainViewController.swift]: Sort please

File: Sasquatch/Sasquatch/ViewControllers/MSMainViewController.swift [hunk 2]
```diff
@@ -73,6 +76,20 @@ class MSMainViewController: UITableViewController, AppCenterProtocol {
...
+    if let msaUserId = UserDefaults.standard.string(forKey: kMSAUserIdKey) {
+        let expiresIn = UserDefaults.standard.integer(forKey: kMSAExpirationDateKey)
+        let expiresDate = Date(timeIntervalSince1970: TimeInterval(exactly:expiresIn)!)
+        let isExpired = Date() > expiresDate
+        if (isExpired) {
+            UserDefaults.standard.removeObject(forKey: kMSAUserIdKey)
+            UserDefaults.standard.removeObject(forKey: kMSAExpirationDateKey)
+        } else {
+            for target in MSTransmissionTargets.shared.transmissionTargets.values {
+                target.propertyConfigurator.setUserId(msaUserId)
```
Comment (guperrot) [inline Sasquatch/Sasquatch/ViewControllers/MSMainViewController.swift]: https://github.com/microsoft/appcenter-sdk-android/pull/1384#pullrequestreview-380607178

File: Sasquatch/Sasquatch/MSAAnalyticsAuthenticationProvider.swift [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+                    if let error = json["error"] as? String, let errorDescription = json["error_description"] as? String {
+                        NSLog("Refresh token error: \"\(error)\": \(errorDescription)")
+                        
+                        // Call the completion handler in the error case to send anonymous logs.
+                        completionHandler(nil, nil)
+                        return
+                    }
+                    let token = json["access_token"]! as! String
+                    let expiresIn = json["expires_in"]! as! Int64
+                    let userId = json["user_id"]! as! String
```
Comment (MatkovIvan) [inline Sasquatch/Sasquatch/MSAAnalyticsAuthenticationProvider.swift]: let's avoid force unwrapping during json parsing

Comment (annakocheshkova) [reply Sasquatch/Sasquatch/MSAAnalyticsAuthenticationProvider.swift]: This is not my code 😒 
And I guess I can see the intention why it was written. We expect this value to be here in any case. If there are not, it's an error so better fail fast.

File: Sasquatch/Sasquatch/ViewControllers/MSMainViewController.swift [hunk 4]
```diff
@@ -7,6 +7,8 @@ import UIKit
 let kMSDefaultDatabaseSize = 10 * 1024 * 1024
 let acProdLogUrl = "https://in.appcenter.ms"
 let ocProdLogUrl = "https://mobile.events.data.microsoft.com"
+let kMSARefreshTokenKey = "MSARefreshToken"
+let kMSAUserIdKey = "MSAUserId"
```
[the remark this answers is not in the document]
Comment (guperrot) [reply Sasquatch/Sasquatch/ViewControllers/MSMainViewController.swift]: (Also should rename PR title)

File: Sasquatch/Sasquatch/ViewControllers/MSSignInViewController.swift [hunk 5]
```diff
@@ -65,33 +48,35 @@ class MSSignInViewController: UIViewController, WKNavigationDelegate, MSAnalytic
...
       self.webView.load(URLRequest(url: url))
     }
   }
 
   // The sign in flow.
   func signIn(url: URL) {
-    if url.absoluteString.contains((self.baseUrl + self.redirectEndpoint)) {
-      if let newUrl = URL(string: self.baseUrl + self.redirectEndpoint + "?" + url.fragment!) {
+    if url.absoluteString.contains((kMSABaseUrl + kMSARedirectEndpoint)) {
+      if let newUrl = URL(string: kMSABaseUrl + kMSARedirectEndpoint + "?" + url.fragment!) {
```
Comment (guperrot) [inline Sasquatch/Sasquatch/ViewControllers/MSSignInViewController.swift]: Screen Shot 2020-03-26 at 2 43 58 PM

It crashes on simulator iphone XR / XCode 10.1

Comment (annakocheshkova) [reply Sasquatch/Sasquatch/ViewControllers/MSSignInViewController.swift]: Strange, my change just moves constants..

Comment (annakocheshkova) [reply Sasquatch/Sasquatch/ViewControllers/MSSignInViewController.swift]: Ok found what was the case

**Comment by guperrot** [comment]
The scenario I described still does not work even after the last commit.

**Comment by annakocheshkova** [comment]
@guperrot same here, I followed your flow again today on a new simulator (just in case), see in logs:
```
[AppCenter] VERBOSE: -[MSOneCollectorIngestion willSendHTTPRequestToURL:withHeaders:]/144 Headers: Tickets = {"***":"p:***"}, Content-Encoding = gzip, Client-Version = ACS-iOS-ObjectiveC-no-3.1.1-no, apikey = ******************************************************************797-7361,******************************************************************912-6819,******************************************************************137-7039, Content-Type = application/x-json-stream; charset=utf-8, Upload-Time = 1585380802888
```
[Link to the corresponding event](https://portal.aria.ms/event/1aa046cfdc8f49bdbd64190290caf7dd/testeventafterkill/inspector), see the Id

**Comment by guperrot** [comment]
> Tested on a real device and the tokens are sent on new events after app killed. The signing out doesn't work but not a regression it's also happening in develop so I added a bug.

We should address it now by simply removing the persistence for the next restart when signOut clicked. Otherwise our testers will definitely open a bug.

**Comment by guperrot** [comment]
> @guperrot same here, I followed your flow again today on a new simulator (just in case), see in logs:
> 
> ```
> 2020-03-28 10:37:40.020 SasquatchObjC[14338:2493397] [AppCenter] VERBOSE: 
> -[MSOneCollectorIngestion willSendHTTPRequestToURL:withHeaders:]/144 
> Headers: Tickets = {"***":"p:***"}, 
> Content-Encoding = gzip, Client-Version = ACS-iOS-ObjectiveC-no-3.1.1-no, 
> apikey = ******************************************************************797-7361,******************************************************************912-6819,******************************************************************137-7039, 
> Content-Type = application/x-json-stream; charset=utf-8, Upload-Time = 1585381060019
> ```
> 
> [Link to the corresponding event](https://portal.aria.ms/event/1aa046cfdc8f49bdbd64190290caf7dd/testeventafterkill/inspector), see the Id

I was still on develop for apple and my terminal was showing android branch 😅 

It works indeed now.

**Pull request merged**